protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,717 @@
+        if let Poll::Ready(Some((src_peer_id, result))) = self.circuit_deny_futs.poll_next_unpin(cx)
+        {
+            match result {
+                Ok(()) => {
+                    return Poll::Ready(ProtocolsHandlerEvent::Custom(
+                        Event::InboundCircuitReqDenied { src_peer_id },
+                    ))
+                }
+                Err(error) => {
+                    return Poll::Ready(ProtocolsHandlerEvent::Custom(
+                        Event::InboundCircuitReqDenyFailed { src_peer_id, error },
+                    ))
+                }
+            }
+        }
+
+        // Send errors to transport.
+        while let Poll::Ready(Some(())) = self.send_error_futs.poll_next_unpin(cx) {}
+
+        // Check status of lend out substreams.
+        loop {
+            match self.alive_lend_out_substreams.poll_next_unpin(cx) {
+                Poll::Ready(Some(Err(oneshot::Canceled))) => {}
+                Poll::Ready(Some(Ok(v))) => void::unreachable(v),
+                Poll::Ready(None) | Poll::Pending => break,
+            }
+        }
+
+        // Update keep-alive handling.
+        if matches!(self.reservation, Reservation::None)
+            && self.alive_lend_out_substreams.is_empty()
+            && self.circuit_deny_futs.is_empty()
+        {
+            match self.keep_alive {
+                KeepAlive::Yes => {
+                    self.keep_alive = KeepAlive::Until(Instant::now() + Duration::from_secs(10));
+                }
+                KeepAlive::Until(_) => {}
+                KeepAlive::No => panic!("Handler never sets KeepAlive::No."),
+            }
+        } else {
+            self.keep_alive = KeepAlive::Yes;
+        }
+
+        Poll::Pending
+    }
+}
+
+enum Reservation {
+    /// The Rreservation is accepted by the relay.
+    Accepted {
+        /// [`None`] if reservation does not expire.
+        renewal_timeout: Option<Delay>,
+        /// Buffer of messages to be send to the transport listener.
+        pending_msgs: VecDeque<transport::ToListenerMsg>,
+        to_listener: mpsc::Sender<transport::ToListenerMsg>,
+    },
+    /// The reservation is being renewed with the relay.
+    Renewing {
+        /// Buffer of messages to be send to the transport listener.
+        pending_msgs: VecDeque<transport::ToListenerMsg>,
+    },
+    None,
+}
+
+impl Reservation {
+    fn accepted(
+        &mut self,
+        renewal_timeout: Option<Delay>,
+        addrs: Vec<Multiaddr>,
+        to_listener: mpsc::Sender<transport::ToListenerMsg>,
+        local_peer_id: PeerId,
+    ) -> Event {
+        let (renewal, mut pending_msgs) = match std::mem::replace(self, Self::None) {
+            Reservation::Accepted { pending_msgs, .. }
+            | Reservation::Renewing { pending_msgs, .. } => (true, pending_msgs),
+            Reservation::None => (false, VecDeque::new()),
+        };
+
+        pending_msgs.push_back(transport::ToListenerMsg::Reservation(Ok(
```

Comment:
General question:

Is the reason we don't just send this event to the listener right away because we only want to trigger side-effecty things from within the context of a `poll` call? And otherwise we want to just remember we need to do something then actually do it when we are polled.

Reply:
This is a pretty unique situation in rust-libp2p where we could action this right away. Normally, the only way for communication with other components is by returning something from `poll`.

Hence, I didn't even think about this here!

Perhaps it would be better to action these messages right away here?

@mxinden what do you think?

Reply:
> we only want to trigger side-effecty things from within the context of a poll call?

While that might be something worth considering standardizing in the future, in this case the reason is a lot simpler.

To send something to the listener we use an `futures::channel::mpsc::channel`. Sending something into a `futures::channel::mpsc::Sender` requires access to a `Context`. The channel might not have capacity for the additional item. In case it does not, we need the `Context` to be woken up once it does and buffer the item.

Given that the caller of `Reservation::accepted` does not have access to a `Context`, we delay the sending to `Reservation::poll`.

@MarcoPolo does the above make sense?

Reply:
ah yes, of course. I missed that we need the Context here. That makes sense, thanks!